docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,429 @@
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  COMPLETE
+
+-- Restart new version of the application that expects the INET type.
+```
+
+Electing not to use the `DEFERRED` option is equivalent to running a
+`DEFERRED` followed immediately by a `COMPLETE`. We will refer to this
+as "auto-complete" mode.
+
+The `COMPLETE` option requires the entire, original `ALTER TABLE`
+command as a sanity-check. If the type conversion generated any output,
+it will be emitted by this command. If the type conversion has not yet
+completed, this command will block until it has. It is not an error to
+run `COMPLETE` before the conversion job has finished; it is simply a
+signal to indicate that the SQL consumers are prepared to handle the new
+column type.
+
+## Key constraints
+
+Columns that are the target of a primary or foreign key constraint will
+not be supported under this RFC. It could be possible to support foreign
+keys in the future by cascading the type change, but this would greatly
+increase the amount of coordination required to effect the total schema
+change.
+
+### Interleaved tables
+
+Changing the types of primary keys is not within the scope of this RFC,
+so this feature does not interact with interleaved tables. When changes
+to primary keys are supported, a change to the PK of a parent table
+would have to be cascaded to the corresponding PK prefix column of any
+interleaved descendant tables.
+
+# Reference-level explanation
+
+## Process
+
+Conceptually, a column type change can be thought of as creating a
+shadow column of the new type, which captures writes to the original,
+source, column while a backfill process is run. Once the backfill is
+complete, the shadow column is swapped for the original, which is then
+retired. The shadow column will be anonymous and generally invisible to
+the user until it is swapped in. Shadow columns will be created in the
+same column family as the source column.
+
+This list describes the general type-change case, not all of these steps
+are necessarily executed for trivial or validated changes.
+
+* `ALTER TABLE ... COLUMN ... TYPE` is run, resulting in a new shadow
+  column being created in `DELETE_ONLY` mode. The shadow column has a
+  new column id and a reference to the source column
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
+* The shadow column and indexes are upgraded to `WRITE_ONLY` mode.
+* Nodes that write to the source column will synchronously apply the
+  requested type change and duplicate the write to the shadow column and
+  indexes.
+* A backfill process is distributed across all nodes in the cluster
+  which have data for the table. This backfill will populate the shadow
+  column and shadow indexes, inserting data with the source MVCC
+  timestamp. The backfilled sstables will be inserted into the LSM such
+  that any shadowed write will have priority.
```

Comment:
you can say here that backfilling this column is similar to how we backfill a computed column today